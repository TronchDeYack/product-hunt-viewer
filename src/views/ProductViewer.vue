<template>
  <div class="product-viewer">
    <DaysSelector />
    <Stats :posts="postsNbr" :votes="votesNbr" :comments="commentsNbr" :makers="makersNbr" />
    <Posts :posts="posts"/>
  </div>
</template>

<style lang="scss">
.product-viewer {
  .title {
    color: red;
    margin-bottom: 0;
  }
  .subtitle {
    font-size: 12px;
    font-weight: lighter;
    margin-top: 0;
    margin-bottom: 2rem;
  }

  .posts {
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    width: 80%;
    margin: 0 auto;
    justify-content: center;
  }
}
</style>

<script>
import { mapState } from 'vuex';
import gql from 'graphql-tag';

import Posts from '@/components/Posts.vue';
import Stats from '@/components/Stats.vue';
import DaysSelector from '@/components/DaysSelector.vue';
import { getLowBoundary, getHighBoundary } from '@/utils/date';

export default {
  name: 'product-viewer',
  components: {
    DaysSelector,
    Stats,
    Posts,
  },
  apollo: {
    posts: {
      query: gql`query Posts($postedAfter: DateTime, $postedBefore: DateTime) {
        posts(postedAfter: $postedAfter, postedBefore: $postedBefore, ) {
          edges {
            node {
              id
              name
              description
              createdAt
              votesCount
              makers {
                id
              }
              commentsCount
              thumbnail {
                url
              }
            }
          }
        }
      }`,
      variables() {
        return {
          postedAfter: getLowBoundary(this.day),
          postedBefore: getHighBoundary(this.day),
        };
      },
    },
  },
  computed: {
    ...mapState({
      day: state => state.day,
    }),
    postsNbr() {
      return this.posts ? this.posts.edges.length : 0;
    },
    votesNbr() {
      return this.posts ? this.posts.edges.reduce(
        (accumulator, currentEdge) => accumulator + currentEdge.node.votesCount, 0,
      ) : 0;
    },
    commentsNbr() {
      return this.posts ? this.posts.edges.reduce(
        (accumulator, currentEdge) => accumulator + currentEdge.node.commentsCount, 0,
      ) : 0;
    },
    makersNbr() {
      if (!this.posts) {
        return 0;
      }
      const idMakers = this.posts.edges.reduce(
        (accumulator, currentEdge) => accumulator.concat(
          currentEdge.node.makers.map(maker => maker.id),
        ), [],
      );
      return new Set(idMakers).size;
    },
  },
  data() {
    return {
      posts: null,
    };
  },
  watch: {
    day() {
      this.$apollo.queries.posts.refresh(this.day);
    },
  },
};
</script>
