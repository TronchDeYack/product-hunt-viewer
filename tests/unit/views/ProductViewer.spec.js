import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Vue from 'vue';

import ProductViewer from '@/views/ProductViewer.vue';
import initializeStore from '../utils';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('ProductViewer', () => {
  let wrapper;

  const day = 15;

  const posts = {
    edges: [{
      node: {
        id: '178112',
        name: 'Mailmeteor',
        description: 'Mailmeteor is the easiest and privacy-first mail merge add-on for Gmail. ðŸŒ Email a lot of people quickly ðŸŒ Real-time tracking, attachments, alias... ðŸŒ Increase your answer rate dramatically ðŸ’« Send 100 emails a day for free',
        createdAt: '2019-12-19T08:00:00Z',
        votesCount: 212,
        makers: [{
          id: '1',
        }, {
          id: '2',
        }],
        commentsCount: 150,
        thumbnail: {
          url: 'https://ph-files.imgix.net/4d44f9de-3d83-4a29-9285-a0651bc948e0?auto=format&fit=crop',
        },
      },
    }, {
      node: {
        id: '178447',
        name: 'Slapdash 1.0 ',
        description: 'A rethink of how we work, with a focus on speed, ergonomics and fun. It\'s the missing operating system for your cloud apps.',
        createdAt: '2019-12-19T08:03:00Z',
        votesCount: 207,
        makers: [{
          id: '2',
        }, {
          id: '3',
        }, {
          id: '4',
        }],
        commentsCount: 250,
        thumbnail: {
          url: 'https://ph-files.imgix.net/42122d73-bde8-49af-99bd-67d9351f978f?auto=format&fit=crop',
        },
      },
    }],
  };

  beforeEach(() => {
    wrapper = shallowMount(ProductViewer, {
      localVue,
      mocks: {
        $store: initializeStore(day, jest.fn()),
      },
    });
  });

  it('Should render.', () => {
    wrapper.setData({
      posts,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('Should initialize "posts" data to null.', () => {
    expect(wrapper.vm.$data.posts).toBeNull();
  });

  it('Should recall "posts" query when the selected day is updated.', (done) => {
    const refreshMock = jest.fn();
    const localStore = initializeStore(day, jest.fn());

    wrapper = shallowMount(ProductViewer, {
      localVue,
      mocks: {
        $store: localStore,
        $apollo: {
          queries: {
            posts: {
              refresh: refreshMock,
            },
          },
        },
      },
    });

    localStore.state.day = 25;

    Vue.nextTick(() => {
      expect(refreshMock).toHaveBeenCalledWith(25);
      done();
    });
  });

  describe('Computed', () => {
    describe('Posts loaded', () => {
      beforeEach(() => {
        wrapper.setData({
          posts,
        });
      });

      it('Should compute the numbers of posts based on the results.', () => {
        expect(wrapper.vm.postsNbr).toBe(2);
      });

      it('Should compute the numbers of comments based on the results.', () => {
        expect(wrapper.vm.commentsNbr).toBe(400);
      });

      it('Should compute the numbers of votes based on the results.', () => {
        expect(wrapper.vm.votesNbr).toBe(419);
      });

      it('Should compute the numbers of makers by deleting clone based on the results.', () => {
        expect(wrapper.vm.makersNbr).toBe(4);
      });
    });

    describe('Posts loading', () => {
      it('Should compute the numbers of posts with 0 when there is no post yet.', () => {
        expect(wrapper.vm.postsNbr).toBe(0);
      });

      it('Should compute the numbers of comments with 0 when there is no post yet.', () => {
        expect(wrapper.vm.commentsNbr).toBe(0);
      });

      it('Should compute the numbers of votes with 0 when there is no post yet.', () => {
        expect(wrapper.vm.votesNbr).toBe(0);
      });

      it('Should compute the numbers of makers by deleting clone with 0 when there is no post yet.', () => {
        expect(wrapper.vm.makersNbr).toBe(0);
      });
    });

    it('Should compute day from the store.', () => {
      expect(wrapper.vm.day).toBe(day);
    });
  });
});
