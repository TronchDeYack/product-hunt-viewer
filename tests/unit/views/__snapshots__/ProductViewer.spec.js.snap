// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductViewer Should render. 1`] = `
<div class="product-viewer">
  <daysselector-stub></daysselector-stub>
  <stats-stub posts="2" votes="419" comments="400" makers="4"></stats-stub>
  <posts-stub posts="[object Object]"></posts-stub>
</div>
`;
